File: SwiftAlgorithm/BOJ/20000/20922.swift
//
//  20922.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/03/29.
//

import Foundation

let nk = readLine()!.components(separatedBy: " ").map { Int($0)! }
let n = nk[0], k = nk[1]
let input = readLine()!.components(separatedBy: " ").map { Int($0)! }

var dict: [Int: Int] = [:]

var result = 0
var left = 0, right = 0
while right < n {
    guard let count = dict[input[right]] else {
        dict[input[right]] = 1
        right += 1
        result = max(result, right - left)
        continue
    }
    
    //넣을 수 있음
    if count < k {
        dict[input[right]] = count + 1
        right += 1
    } else { //못 넣음
        dict[input[left]]! -= 1
        if dict[input[left]]! == 0 {
            dict[input[left]] = nil
        }
        
        left += 1
    }
    
    result = max(result, right - left)
}

print(result)

File: Python/BOJ/투포인터/20922.py
//
//  20922.py
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/06/12.
//

import sys; readLine = sys.stdin.readline

# right를 움직이면서 arr 원소가 k 를 넘는지 확인
# k를 넘으면 arr 원소가 k를 넘지 않을 때까지 left를 움직임
# 반복 마지막에 result를 max length로 업데이트

n, k = map(int, readLine().split())
arr = list(map(int, readLine().split()))

left = 0; right = 0
checked = [0 for _ in range(100001)]
checked[arr[left]] = 1
result = 1

while left <= right and right < n:
    right += 1
    if right == n:
        continue
    checked[arr[right]] += 1

    while checked[arr[right]] > k:
        checked[arr[left]] -= 1
        left += 1

    result = max(result, right - left + 1)

print(result)
